03_Fiches_thematiques/Fiche_duckdb.qmd

```diff
@@ -0,0 +1,522 @@
+
+
+## Connexion à une base de données
+
+**`duckdb` est une base de données distante et s'utilise comme telle: il faut ouvrir une connexion, puis "charger" les données dans la base de données pour les manipuler.** 
+
+Comme beaucoup d'autres bases de données (distantes ou locales), on ouvre une connexion au moteur `duckdb` avec une base de données en mémoire vive de la façon suivante :
+```{r}
+con <- DBI::dbConnect(duckdb::duckdb())
+```
+
+Concrètement, cette commande crée une nouvelle base de données `duckdb` dans la mémoire vive. Cette base de données ne contient aucune donnée lorsqu'elle est créée.
+
+À la fin du traitement ou du programme, on ferme la connexion avec le code ci-dessous. L'option `shutdown` est importante : elle permet de fermer complètement la session `duckdb` et de libérer la mémoire utilisée. Si on n'utilise pas cette option, il arrive souvent d'avoir des connexions à moitié ouvertes, et de devoir relancer la session `R`.
+```{r}
+DBI::dbDisconnect(con, shutdown = TRUE)
+```
+
+Pour la suite, on supposera que la connexion est ouverte.
+```{r}
+con <- DBI::dbConnect(duckdb::duckdb()); 
+```
+
+
+
+## Chargement des données et écriture au format Parquet
+
+
+### Déclaration de données provenant de `R`
+
+On charge des données dans `duckdb` au travers de la connexion `con` avec la fonction `duckdb_register()`. Cette methode a l'avantage de ne pas _recopier_ les données. D'autres moyens sont possibles, notamment en passant par `arrow`. L'objet `con` apparaît dans l'onglet `Data` de l'environnement `RStudio`, mais la liste des tables n'y est pas directement accessibles. Pour plus d'informations, se reporter à la documentation du package `DBI`.
+
+```{r}
+con %>% duckdb::duckdb_register(
+  name = "bpe_ens_2018_duckdb", 
+  df = doremifasolData::bpe_ens_2018 |> as_tibble())
+```
+
+Le code ci-dessous permet de vérifier que le chargement des données a bien fonctionné. La fonction `tbl` permet d'accéder à un objet de la base de données par le nom (de la table), ou par du code SQL (utilisation avancée). Par défaut, `duckdb` affiche les 10 premières lignes du résultat, sans effectuer tout le calcul. C'est super pratique !
+```{r}
+con %>% tbl("bpe_ens_2018_duckdb")
+```
+
+
+### Écriture au format parquet 
+
+**Pour écrire une table (ou n'importe quelle requête) sur le disque au format parquet, on suggère d'utiliser la librairie `arrow`.**
+```{r}
+con %>% tbl("bpe_ens_2018_duckdb") %>% 
+  arrow::to_arrow() %>% arrow::write_dataset("bpe_ens_2018_dataset")
+list.files("bpe_ens_2018_dataset") # liste des fichiers du répertoire bpe_ens_2018_dataset/
+```
+
+Une autre possibilité est de passer par la commande SQL `COPY ... TO`. Dans mon expérience, passer par `arrow` permet d'exécuter des calculs plus complexes sans faire déborder la mémoire (`arrow` sauvegarde au fur et à mesure), et la syntaxe est plus facile à manipuler, notamment quand on souhaite ajouter des options.
```

Comment:
```suggestion
Une autre possibilité est de passer par la commande SQL `COPY ... TO`. D'expérience, passer par `arrow` permet d'exécuter des calculs plus complexes sans faire déborder la mémoire (`arrow` sauvegarde au fur et à mesure), et la syntaxe est plus facile à manipuler, notamment quand on souhaite ajouter des options.
```

Reply:
D'après mes tests, le `COPY ... TO` était un peu plus efficace que le `arrow::write_dataset`, il sauvegarde aussi au fur et à mesure :
https://nbc.github.io/notes-en-vrac/cas-pratique-traitement-duckdb.html#la-solution-arrowwrite_dataset

Reply:
Je n'ai jamais réussi à faire marcher `COPY .. TO` avec une sauvegarde au fur et à mesure. 
Dans mes benchmarks (sur fichier moyen), il me faut 500Mo de RAM en passant par `COPY ... TO` avec `per_thread_output`, mais seulement 108Mo avec `write_dataset`